--- sys/dev/usb/input/uhid.c.orig	2022-08-02 16:28:32.777495000 +0200
+++ sys/dev/usb/input/uhid.c	2022-08-02 18:15:54.000000000 +0200
@@ -218,6 +218,12 @@
 				actlen = sc->sc_isize;
 			usb_fifo_put_data(sc->sc_fifo.fp[USB_FIFO_RX], pc,
 			    0, actlen, 1);
+
+			/*
+			 * Do not do read-ahead, because this may lead
+			 * to data loss!
+			 */
+			return;
 		} else {
 			/* ignore it */
 			DPRINTF("ignored transfer, %d bytes\n", actlen);
