--- sys/dev/igc/igc_i225.c.orig	2023-02-13 12:48:18.499154000 +0100
+++ sys/dev/igc/igc_i225.c	2023-02-25 16:55:18.000000000 +0100
@@ -172,14 +172,12 @@
 	/* Verify phy id and set remaining function pointers */
 	switch (phy->id) {
 	case I225_I_PHY_ID:
+	default:
 		phy->type		= igc_phy_i225;
 		phy->ops.set_d0_lplu_state = igc_set_d0_lplu_state_i225;
 		phy->ops.set_d3_lplu_state = igc_set_d3_lplu_state_i225;
 		/* TODO - complete with GPY PHY information */
 		break;
-	default:
-		ret_val = -IGC_ERR_PHY;
-		goto out;
 	}
 
 out:
